pyfar/signals/deterministic.py

```diff
@@ -256,6 +389,478 @@ def exponential_sweep_time(n_samples, frequency_range, n_fade_out=90,
+        n_samples=n_samples,
+        sweep_type='exponential',
+        frequency_range=frequency_range,
+        butterworth_order=butterworth_order,
+        start_margin=start_margin,
+        stop_margin=stop_margin,
+        fade_in=fade_in,
+        fade_out=fade_out,
+        sampling_rate=sampling_rate)
+
+    return signal, group_delay
+
+
+def magnitude_spectrum_weighted_sweep(
+        n_samples, magnitude_spectrum, start_margin, stop_margin,
+        fade_in=None, fade_out=None, sampling_rate=44100):
+    """
+    Generate sine sweep with arbitrary magnitude spectrum in the frequency
+    domain.
+
+    Sine sweep synthesis according to [#]_. Frequency domain synthesis
+    exhibits smooth magnitude spectra in trade of a slightly irregular temporal
+    envelope. Time domain synthesis exhibits a constant temporal envelope in
+    trade of slight ripples in the magnitude response. But there is currently
+    no method to design sine sweeps with arbitrary magnitude response in the
+    time domain.
+
+    .. note::
+        The envelope of the sweep time signal should be constant, apart from
+        slight overshoots at the beginning and end. If this is not the case,
+        try to increase `n_samples`, `start_margin`, `stop_margin`, `fade_in`
+        or `fade_out`, or provide a more smooth magnitude spectrum.
+
+    Parameters
+    ----------
+    n_samples : int
+        The length of the sweep in samples.
+    magnitude_spectrum : Signal
+        The magnitude spectrum as a Signal object. Must be at least `n_samples`
+        long and is zero-padded to `n_samples` otherwise.
+    start_margin : int, float
+        The time in samples, at which the sweep starts. The start margin is
+        required because the frequency domain sweep synthesis has pre-ringing
+        in the time domain.
+    stop_margin : int, float
+        Time in samples, at which the sweep stops. This is relative to
+        `n_samples`, e.g., a stop margin of 100 samples means that the sweep
+        ends at sample ``n_samples-100``. This is required, because the
+        frequency domain sweep synthesis has post-ringing in the time domain.
+    fade_in : int
+        Duration of a squared sine fade-in in samples. The fade starts at the
+        first sample of the sweep that is closer than 60 dB to the absolute
+        maximum of the sweep time signal. The default ``0`` does not apply
+        a fade-in.
+    fade_out : int
+        Duration of a squared cosine fade-out in samples. The fade ends at the
+        last sample of the sweep that is closer than 60 dB to the absolute
+        maximum of the sweep time signal. The default ``0`` does not apply
+        a fade-out.
+    sampling_rate : int
+        The sampling rate in Hz. The default is ``44100``.
+
+    Returns
+    -------
+    sweep : Signal
+        The sweep signal. The Signal is in the time domain, has a maximum
+        absolute amplitude of 1 and the ``rms`` FFT normalization
```

Comment:
I think `rms` normalization is not correct here

Reply:
will be corrected